### ISSUE TITLE
Page cleaning destroys format 

### ISSUE THREAD

**Issue by elben10** [pymupdf/PyMuPDF#519]
## Describe the bug
`apply_redactions` changes the structure of other blocks in the document. For instance take a look at sidebar. 

## To Reproduce
The pdf is available at https://www.ft.dk/samling/20191/spoergsmaal/s1150/svar/1659504/2191260.pdf
```python
import fitz

doc = fitz.open("2191260.pdf")
page = doc[0]
page.addRedactAnnot((0, 0, 10, 10), "??")
page.apply_redactions()
doc.save("transformed.pdf")
```
## Expected behavior
The expected behaviour is that the document would have "??" added in the upper left corner, and the rest of the document would appear as is.

## Screenshots
![](https://i.ibb.co/6BMH51B/Sk-rmbillede-2020-05-27-kl-09-03-52.png)
 
## Your configuration (mandatory)
```
3.8.2 (v3.8.2:7b3ab5921f, Feb 24 2020, 17:52:18)
[Clang 6.0 (clang-600.0.57)]
 darwin

PyMuPDF 1.16.18: Python bindings for the MuPDF 1.16.0 library.
Version date: 2020-04-22 14:00:00.
Built for Python 3.8 on darwin (64-bit).
```
Installed using `pip install PyMuPDF`

**Comment by JorjMcKie** [comment]
Checked this out:
This has nothing to do with redactions actually. The same destruction happens if you clean the page of the original document: ``page.cleanContetns()`` - or even ``mutool clean -s 2191260.pdf``.
Applying redactions internally invokes page cleaning, which then produces this.

This is an upstream bug in MuPDF. You can report it to their bug tracking system together with your original file (without the redaction annot).
Or I can do it with your permission.

**Comment by elben10** [comment]
I can only replicate the behaviour when using `page.cleanContents()`. It destroys the format.
If I use `mutool clean 2191260.pdf  processed.pdf` the document doesn't change.

**Comment by JorjMcKie** [comment]
You forgot the ``-s`` parameter: ``mutool clean -s ...``

**Comment by JorjMcKie** [comment]
Here is the bug system of Artifex/MuPDF: https://bugs.ghostscript.com/enter_bug.cgi.

**Comment by JorjMcKie** [comment]
Just to prevent a potentially wrong impression:
Cleaning pages' ``/Contents`` objects normally works fine. It does a lot of things:
* Syntactical cleaning of the postscript-like painting commands which makes up the appearance of the page. Among those actions: standardize number formats, omit duplicate commands,, put each command on its own line, replace multiple white spaces with one normal space, ensure that ``/Contents`` start with command ``q`` and end with command ``Q``. This ensures that establishing any non-standard page geometry remains local: adding additional objects to that page can rely on standard PDF geometry, ...
* In case of several contents objects, unify them into one
* Ensure congruence of resources named in the contents and the ``/Resources`` object of the page (i.e. fonts, images, Form XObjects, etc.)

So your approach does work in the vast majority of cases ,,,

**Comment by elben10** [comment]
The bug is reported and can be found [here](https://bugs.ghostscript.com/show_bug.cgi?id=702441)

**Comment by JorjMcKie** [comment]
This is great! A reaction from a broader public increases the chance for a prompt response from them.

**Comment by JorjMcKie** [comment]
Just to let you know:
The new version 1.18.0 of MuPDF has finally resolved this problem.
I am testing with the release candidate and will publish the resp. new PyMuPDF as soon as an official MuPDF version is out.

**Comment by denyskotsur** [comment]
I'm facing the same issue, please @JorjMcKie keep updated this thread, so we can enjoy the new working version :) 

Thanks!

**Comment by JorjMcKie** [comment]
@denyskotsur - sure!

**Comment by JorjMcKie** [comment]
To whom it may concern: There are [Linux](https://github.com/pymupdf/PyMuPDF-wheels/tree/linux) and [Mac OSX](https://github.com/pymupdf/PyMuPDF-wheels/tree/osx) pre-release wheels downloadable now, which should resolve this issue - and a few others, too.
Please drop me a note if you need a Windows wheel.

**Comment by JorjMcKie** [comment]
This issue is fixed with v1.18.0 - now on PyPI.
